### PR TITLE
Added a few in/out bf16 EPILOGUEs

### PR DESCRIPTION
**Pull request by solaslin**
Resolves [SWDEV-435742](https://ontrack-internal.amd.com/browse/SWDEV-435742)

- [x] **HIPBLASLT_EPILOGUE_BGRADB** (A=BF16 B=BF16 D=BF16 bias=BF16, **NT**) 
  - Added BF16 in SumUnroll part 
- [x] **HIPBLASLT_EPILOGUE_GELU_AUX_(BIAS)** (A=bf16 B=BF16 D=BF16 bias=BF16, **TN**)
  - Support BF16 for E buffer (useE + Gradient=False)
- [x] **HIPBLASLT_EPILOGUE_DGELU** (A=BF16 B=BF16 D=BF16, **NN**)
  - Support BF16 for E buffer (useE + Gradient=True)
- [x] Added examples in clients/samples
- [x] gtest

### CODE REVIEW

File: tensilelite/Tensile/Components/GlobalWriteBatch.py [hunk 1]
```diff
@@ -796,6 +796,11 @@ def _emitNonatomicAdd(self, module: Module):
             packdata = PackData_F16()
             module.add(packdata(self.gwvw, tmpVgpr, vgprDst, tmpVgpr=tmpVgpr, inputPrefix=prefixStr, prefixOffset=prefixOffset))
             vgprDst = tmpVgpr
+          elif self.kernel["ProblemType"]["DataTypeE"].isBFloat16():
+            packdata = PackData_BF16()
+            module.add(packdata(self.gwvw, tmpVgpr, vgprDst, self.cvtVgprStruct, self.tmpS01, self.laneSGPRC,
+                                tmpVgpr=tmpVgpr, inputPrefix=prefixStr, prefixOffset=prefixOffset))
+            vgprDst = tmpVgpr
           elif self.kernel["ProblemType"]["DataTypeE"].isSingle():
```
Comment (solaslin) [inline tensilelite/Tensile/Components/GlobalWriteBatch.py:804]: Added BF16 part for GELU AUX (UseE and **NOT** Gradient)

File: tensilelite/Tensile/Components/GlobalWriteBatch.py [hunk 2]
```diff
@@ -818,6 +823,12 @@ def _emitNonatomicAdd(self, module: Module):
               dataEV2 = dataE + vi // 2
               selectbit = SelectBit.WORD_0 if (self.gwvw != 1 and vi % 2 == 0) or (self.gwvw == 1 and elementIdx % 2 == 0) else SelectBit.WORD_1
               gradientCvtModule.add(VCvtF16toF32(dst=vgpr(dataEV), src=vgpr(dataEV2+loadOffset), sdwa=SDWAModifiers(src0_sel=selectbit), comment="gwvw %d, elementIdx %d"%(self.gwvw, elementIdx)))
+          elif activationCDataType.isSingle() and self.kernel["ProblemType"]["DataTypeE"].isBFloat16():
+            for vi in range(0, self.gwvw):
+              dataEV  = dataE + vi
+              dataEV2 = dataE + vi // 2
+              selectWord = 0 if (self.gwvw != 1 and vi % 2 == 0) or (self.gwvw == 1 and elementIdx % 2 == 0) else 1
+              module.add(VCvtBF16toFP32(dst=(dataEV), src=(dataEV2+loadOffset), vgprMask=(self.cvtVgprStruct.vgprBf16Mask), vi=(selectWord), additionalCmts="gwvw %d, elementIdx %d"%(self.gwvw, elementIdx)))
           else:
```
Comment (solaslin) [inline tensilelite/Tensile/Components/GlobalWriteBatch.py:832]: Added BF16 part for DGELU (UseE and Gradient)

File: tensilelite/Tensile/Components/GlobalWriteBatch.py [hunk 3]
```diff
@@ -1418,10 +1429,7 @@ def _addSumAlphaWithCBeta(self, kernel, ss, gwvw, elementIdx, vc0, tmpVgpr, cvtV
           # src1 = dataV = f16.lo = opsel 10 or 11 depending on even/odd
           # src2 = sumIdxV = f32 = opsel 00
           dataCExternal = ss.elementData[elementIdx] + vi//2
-          if (vi%2) == 1:
-            module.add(VAndB32(dst=vgpr(tmpVgpr), src0=vgpr(dataCExternal), src1=vgpr(cvtVgprStruct.vgprBf16Mask), comment="convert bf16 to fp32"))
-          else:
-            module.add(VLShiftLeftB32(dst=vgpr(tmpVgpr), shiftHex=16, src=vgpr(dataCExternal), comment="convert bf16 to fp32" ))
+          module.add(VCvtBF16toFP32(dst=(tmpVgpr), src=(dataCExternal), vgprMask=(cvtVgprStruct.vgprBf16Mask), vi=(vi)))
           newSumIdxV = sumIdxV - self.parentWriter.states.c.startVgprValu
```
Comment (solaslin) [inline tensilelite/Tensile/Components/GlobalWriteBatch.py:1433]: Not directly related to this PR, but I found these four lines can be replaced with the ExtInstruction

File: tensilelite/Tensile/Components/PackData.py [hunk 4]
```diff
@@ -77,22 +77,29 @@ def __call__(self, gwvw, destIdx, elementSumIdx, bf16CVTVgprStruct, tmpS01, lane
...
-            module.add(VCndMaskB32(dst=vgpr(formatVgpr), src0=vgpr(vgprBf16Temp), src1=vgpr(vgprFp32Nan), src2=sgpr(tmpS01,laneSGPRC)))
+            module.add(VCndMaskB32(dst=vgpr(tmpDst), src0=vgpr(vgprBf16Temp), src1=vgpr(vgprFp32Nan), src2=sgpr(tmpS01,laneSGPRC)))
             if vi%2 == 0:
-                module.add(VLShiftRightB32(dst=vgpr(formatVgpr), shiftHex=16, src=vgpr(formatVgpr), comment="convert C to bf16"))
+                module.add(VLShiftRightB32(dst=vgpr(tmpDst), shiftHex=16, src=vgpr(tmpDst), comment="convert C to bf16"))
             elif vi%2 == 1:
                 d = destIdx + vi//2
-                module.add(VAndOrB32(dst=vgpr(d), src0=vgpr(formatVgpr), src1=vgpr(vgprBf16Mask), src2=vgpr(formatting(sumIdxV-1, inputPrefix, prefixOffset)), comment="pack two bf16 to dword"))
+                module.add(VAndOrB32(dst=vgpr(d), src0=vgpr(tmpDst), src1=vgpr(vgprBf16Mask), src2=vgpr(tmpDst_1), comment="pack two bf16 to dword"))
         return module
```
Comment (solaslin) [inline tensilelite/Tensile/Components/PackData.py:103]: PackData_BF16: added a tmpVgpr for buffer E, same idea as half

File: tensilelite/Tensile/Components/SumUnroll.py [hunk 5]
```diff
@@ -109,6 +114,18 @@ def loopSum(self, writer, kernel, tP, u, innerUnroll):
...
+                    if vgprPerInput > 1 and (vgprPerInput % 2 == 0):
+                        for inputIdx in range(0, vgprPerInput):
+                            imod.add(VCvtBF16toFP32(dst=(tmpVgpr), src=("%s+%s"%(valuStr, iui_new_offset + inputIdx)), vgprMask=None, vi=0))
+                            imod.add(VCvtBF16toFP32(dst=(tmpVgpr+1), src=("%s+%s"%(valuStr, iui_new_offset + inputIdx)), vgprMask=hiBitsMaskVgpr, vi=1))
+                            imod.add(VAddF32(dst=vgpr(valuSumStr), src0=vgpr(tmpVgpr), src1=vgpr(valuSumStr), comment="sum K"))
+                            imod.add(VAddF32(dst=vgpr(valuSumStr), src0=vgpr(tmpVgpr+1), src1=vgpr(valuSumStr), comment="sum K"))
+                    else:
+                        printExit("Currently unsupported vgprPerInput %u"%vgprPerInput)
+                    writer.vgprPool.checkIn(tmpVgpr)
                 elif (kernel["ProblemType"]["DataType"].isFloat8A() and tc == "A") or \
```
Comment (solaslin) [inline tensilelite/Tensile/Components/SumUnroll.py:129]: Added BF16 part for BGRADA, BGRADB, (For BiasSrcA, B)

**Review by KKyang — APPROVED** [review]

File: clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
+/*******************************************************************************
```
Comment (jichangjichang) [inline clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp]: file name should be more precise.
- this sample use ext api, you should use _ext postfix
- hipblaslt has many epilogue operations, this one is a sample of BGRADB

File: clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    gemmEpilogue.mode           = epilogue;
+    gemmEpilogue.bias_data_type = HIP_R_16BF;
+
+    hipblaslt_ext::GemmInputs inputs;
+    inputs.a     = d_a;
+    inputs.b     = d_b;
+    inputs.c     = d_c;
+    inputs.d     = d_d;
+    inputs.alpha = &alpha;
+    inputs.beta  = &beta;
```
Comment (jichangjichang) [inline clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp]: where is bias pointer??

File: clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp (hunk 6 above)
Comment (solaslin) [reply clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp]: OK, then I'll change name to _gemm_ext_bgradb

File: clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp (hunk 7 above)
Comment (solaslin) [reply clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp]: Oh I misses that, but it seems "rocblaslt_epilogue_valid_args()" doesn't catch the error. Let me add the check in that functions too.

Comment (solaslin) [reply clients/samples/gemm/sample_hipblaslt_gemm_epilogue.cpp]: done

File: library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp [hunk 8]
```diff
@@ -211,10 +211,13 @@ inline rocblaslt_status rocblaslt_epilogue_valid_args(const rocblaslt_epilogue&
...
 
+    bias = nullptr;
     if(is_bias_enabled(epilogue))
+    {
+        if(original_bias == nullptr)
+            status = rocblaslt_status_invalid_pointer;
         bias = (void*)original_bias;
-    else
-        bias = nullptr;
+    }
```
Comment (solaslin) [inline library/src/amd_detail/rocblaslt/src/include/rocblaslt_mat_utils.hpp:221]: need to check if bias ptr is set

File: clients/samples/gemm/sample_hipblaslt_gemm_ext_bgradb.cpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    hipblaslt_ext::GemmInputs inputs;
+    inputs.a     = d_a;
+    inputs.b     = d_b;
+    inputs.c     = d_c;
+    inputs.d     = d_d;
+    inputs.bias  = d_biasVec;
+    inputs.alpha = &alpha;
+    inputs.beta  = &beta;
+    gemm.setProblem(m, n, k, batch_count, gemmEpilogue, inputs);
+
```
Comment (solaslin) [inline clients/samples/gemm/sample_hipblaslt_gemm_ext_bgradb.cpp:130]: set bias pointer

File: clients/samples/gemm/sample_hipblaslt_gemm_ext_bgradb.cpp [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+     *  This is a NN example with epilogue HIPBLASLT_EPILOGUE_BGRADB
+     *  A,B,C,D and Bias are all bf16
+     */
+    Runner<hipblasLtBfloat16, hipblasLtBfloat16, hipblasLtBfloat16, float, float> runner(
+        1024, 512, 1024, 1, 1.f, 1.f, 32 * 1024 * 1024);
+
+    runner.setBiasInfo(true, 'B');
+
+    runner.run([&runner] {
+        simpleGemmExtEpilogue(runner.handle,
```
Comment (jichangjichang) [inline clients/samples/gemm/sample_hipblaslt_gemm_ext_bgradb.cpp]: For the naming of this function, please refer to "simpleGemmAlphaVecExt". It is better to align the same naming rule even this is just a sample.

**Review by jichangjichang — APPROVED** [review]
LGTM. This is for backward pass, please do get Yangwen's apporval for this PR.

File: clients/samples/gemm/sample_hipblaslt_gemm_ext_bgradb.cpp (hunk 10 above)
Comment (solaslin) [reply clients/samples/gemm/sample_hipblaslt_gemm_ext_bgradb.cpp]: done, I rename it to simpleGemmEpilogueBiasVecExt()

**Review by KKyang — APPROVED** [review]

**Pull request merged**